### PR TITLE
libxml2: 2.9.12 → 2.9.13

### PR DESCRIPTION
**Pull request by jtojnar**
###### Motivation for this change

Fix CVE-2022-23308

https://mail.gnome.org/archives/xml/2022-February/msg00009.html

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/libxml2/default.nix [hunk 1]
```diff
@@ -28,34 +43,33 @@ stdenv.mkDerivation rec {
     #   https://github.com/NixOS/nixpkgs/pull/63174
     #   https://github.com/NixOS/nixpkgs/pull/72342
     ./utf8-xmlErrorFuncHandler.patch
```
Comment (jtojnar) [inline pkgs/development/libraries/libxml2/default.nix:45]: This patch might potentially not be needed any more. But we should probably focus on getting the security fixes in first.

**Comment by jtojnar** [comment]
Currently failing in `buildPhase` on some libtool weirdness, maybe we should just switch to CMake for now (but that might cause other problems or dependency cycle):

```
  CCLD     libxml2mod.la
libtool:   error: only absolute run-paths are allowed
make[4]: *** [Makefile:602: libxml2mod.la] Error 1
make[4]: Leaving directory '/build/libxml2-2.9.13/python'
```

File: pkgs/development/libraries/libxml2/default.nix [hunk 2]
```diff
@@ -65,25 +79,29 @@ stdenv.mkDerivation rec {
     (lib.withFeatureAs pythonSupport "python" python)
   ];
 
+  installFlags = lib.optionals pythonSupport [
+    "pythondir=\"${placeholder "py"}/lib/${python.libPrefix}/site-packages\""
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libxml2/default.nix:83]:
```suggestion
    "pythondir=\"${placeholder "py"}/${python.sitePackages}/\""
```

File: pkgs/development/libraries/libxml2/default.nix [hunk 3]
```diff
@@ -1,20 +1,35 @@
...
+, libintl
+, python
+, gettext
+, ncurses
+, findXMLCatalogs
 , libiconv
 , pythonSupport ? enableShared && stdenv.buildPlatform == stdenv.hostPlatform
-, icuSupport ? false, icu ? null
+, icuSupport ? false
+, icu ? null
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libxml2/default.nix:14]:
```suggestion
, icu
```
legacy from before callPackage.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by mweinelt** [comment]
libxslt conflicts due to https://github.com/NixOS/nixpkgs/pull/160470.

**Comment by jtojnar** [comment]
Resolved. I could not reproduce the [previous eval failure](https://gist.github.com/GrahamcOfBorg/6c97791bbb366e24aeed82d47f9af910) from `perlPackages.XMLParser` (maybe it was Darwin bootstrap failure?). Could not find anything relevant on https://github.com/toddr/XML-Parser either.

**Comment by mweinelt** [comment]
It was related to expat and fixed in its 2.4.6 release.

https://github.com/NixOS/nixpkgs/pull/160826#issuecomment-1046074523

**Comment by vcunat** [comment]
OfBorg shows `libxslt` build broken on `aarch64-darwin`.  (For `x86_64-darwin` we don't know, as it timed out earlier in bootstrapping.)

**Comment by jtojnar** [comment]
Anyone from @NixOS/darwin-maintainers able to debug this?

I would guess either https://gitlab.gnome.org/GNOME/libxslt/-/commit/a2db8da1ac30d744eaa38e574bad9841af1d9b5d or https://gitlab.gnome.org/GNOME/libxslt/-/commit/b3076bccdb818caeea427d3e8fa6b148a77ae78a.

Though, I wonder how could it have ever worked when Python 3 support was only added in these commits.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
It would be really nice if we could get the darwin build solved this week before merging this.

**Comment by SuperSandro2000** [comment]
It would be really nice if we could get the darwin build solved this week before merging this.

**Comment by mweinelt** [comment]
Approved once when eval was failing, approved another time after an aarch64-darwin regression is discovered. I don't understand your reasoning.

**Comment by thefloweringash** [comment]
I didn't look very closely at the changes here, but the following is enough for aarch64-darwin:

```diff
diff --git a/pkgs/development/libraries/libxslt/default.nix b/pkgs/development/libraries/libxslt/default.nix
index b12d084e250..d26986a27b9 100644
--- a/pkgs/development/libraries/libxslt/default.nix
+++ b/pkgs/development/libraries/libxslt/default.nix
@@ -2,6 +2,7 @@
 , stdenv
 , fetchurl
 , pkg-config
+, autoreconfHook
 , libxml2
 , findXMLCatalogs
 , gettext
@@ -26,6 +27,8 @@ stdenv.mkDerivation rec {
 
   nativeBuildInputs = [
     pkg-config
+  ] ++ lib.optional (stdenv.isDarwin && stdenv.isAarch64) [
+    autoreconfHook
   ];
 
   buildInputs = [
```

I assume it's another case of requiring a [libtool fix](https://git.savannah.gnu.org/cgit/libtool.git/commit/?id=9e8c882517082fe5755f2524d23efb02f1522490)

**Comment by vcunat** [comment]
Did you have any plans for 21.11?  On a quick glance I'm not so sure about backporting the large `*.nix` diffs.

**Comment by jtojnar** [comment]
No plans. It is pretty problematic – there is quite a few security fixes to cherry-pick them, but there is also too many other changes to bump the package :woman_shrugging:

**Comment by vcunat** [comment]
Well, taking all might be simplest, also in case of future security fixes for 21.11.

**Comment by risicle** [comment]
It does at least look like they're not trying to sneak features into the release - other changes do all look like small bugfixes, some regression fixes.

**Comment by vcunat** [comment]
The libxslt update broke some libvirt docs; I don't know who's at fault but just reverting the libxslt changes from this PR fixed the build for me.  https://hydra.nixos.org/log/vmjxsi83lp9xp4dacgvr4f0f2fd2sklg-libvirt-7.10.0.drv

**Comment by jtojnar** [comment]
Should be fixed by 043f2055b45.

**Comment by nh2** [comment]
21.11 backport of security fix commits: PR https://github.com/NixOS/nixpkgs/pull/171516

**Pull request merged**